underfs/obs/src/main/java/alluxio/underfs/oss/OBSOutputStream.java

```diff
@@ -0,0 +1,164 @@
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * Creates a name instance of {@link OBSOutputStream}.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param client the OBS client
+   */
+  public OBSOutputStream(String bucketName, String key, ObsClient client) throws IOException {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(),
+        "Bucket name must not be null or empty.");
+    Preconditions.checkArgument(key != null && !key.isEmpty(),
+        "OBS object key must not be null or empty.");
+    Preconditions.checkArgument(client != null, "ObsClient must not be null.");
+    mBucketName = bucketName;
+    mKey = key;
+    mObsClient = client;
+
+    mFile = new File(PathUtils.concatPath("/tmp", UUID.randomUUID()));
+
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+      mLocalOutputStream =
+          new BufferedOutputStream(new DigestOutputStream(new FileOutputStream(mFile), mHash));
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
+      mHash = null;
+      mLocalOutputStream = new BufferedOutputStream(new FileOutputStream(mFile));
+    }
+  }
+
+  /**
+   * Writes the given bytes to this output stream. Before close, the bytes are all written to local
+   * file.
+   *
+   * @param b the bytes to write
+   */
+  @Override
+  public void write(int b) throws IOException {
+    mLocalOutputStream.write(b);
+  }
+
+  /**
+   * Writes the given byte array to this output stream. Before close, the bytes are all written to
+   * local file.
+   *
+   * @param b the byte array
+   */
+  @Override
+  public void write(byte[] b) throws IOException {
+    mLocalOutputStream.write(b, 0, b.length);
+  }
+
+  /**
+   * Writes the given number of bytes from the given byte array starting at the given offset to this
+   * output stream. Before close, the bytes are all written to local file.
+   *
+   * @param b the byte array
+   * @param off the start offset in the data
+   * @param len the number of bytes to write
+   */
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    mLocalOutputStream.write(b, off, len);
+  }
+
+  /**
+   * Flushes this output stream and forces any buffered output bytes to be written out. Before
+   * close, the data are flushed to local file.
+   */
+  @Override
+  public void flush() throws IOException {
+    mLocalOutputStream.flush();
+  }
+
+  /**
+   * Closes this output stream. When an output stream is closed, the local temporary file is
+   * uploaded to OSS Service. Once the file is uploaded, the temporary file is deleted.
+   */
+  @Override
+  public void close() throws IOException {
+    if (mClosed.getAndSet(true)) {
```

Comment:
does it make sense to log warning if closing an already closed stream?

Reply:
added